src/wayland/data_device/mod.rs

```diff
@@ -316,215 +209,205 @@ pub fn set_data_device_focus(seat: &Seat, client: Option<Client>) {
 ///
 /// You'll receive events generated by the interaction of clients with your
 /// drag'n'drop in the provided callback. See [`ServerDndEvent`] for details about
 /// which events can be generated and what response is expected from you to them.
-pub fn start_dnd<C>(
-    seat: &Seat,
+pub fn start_dnd<D, C>(
+    dh: &mut DisplayHandle<'_>,
+    seat: &Seat<D>,
     serial: Serial,
     start_data: PointerGrabStartData,
     metadata: SourceMetadata,
-    callback: C,
 ) where
-    C: FnMut(ServerDndEvent) + 'static,
+    D: Dispatch<WlDataOffer, UserData = DataOfferUserData>,
+    D: DataDeviceHandler,
+    D: ServerDndGrabHandler,
+    D: 'static,
 {
-    // TODO: same question as in set_data_device_focus
-    seat.user_data().insert_if_missing(|| {
-        RefCell::new(SeatData::new(
-            seat.arc.log.new(o!("smithay_module" => "data_device_mgr")),
-        ))
-    });
+    seat.user_data()
+        .insert_if_missing(|| RefCell::new(SeatData::new()));
     if let Some(pointer) = seat.get_pointer() {
         pointer.set_grab(
-            server_dnd_grab::ServerDnDGrab::new(
-                start_data,
-                metadata,
-                seat.clone(),
-                Rc::new(RefCell::new(callback)),
-            ),
+            dh,
+            server_dnd_grab::ServerDnDGrab::new(start_data, metadata, seat.clone()),
             serial,
             0,
         );
     }
 }
 
-fn implement_ddm<F, C>(
-    ddm: Main<wl_data_device_manager::WlDataDeviceManager>,
-    callback: Rc<RefCell<C>>,
-    action_choice: Rc<RefCell<F>>,
-    log: ::slog::Logger,
-) -> wl_data_device_manager::WlDataDeviceManager
-where
-    F: FnMut(DndAction, DndAction) -> DndAction + 'static,
-    C: FnMut(DataDeviceEvent) + 'static,
-{
-    use self::wl_data_device_manager::Request;
-    ddm.quick_assign(move |_ddm, req, _data| match req {
-        Request::CreateDataSource { id } => {
-            self::data_source::implement_data_source(id);
-        }
-        Request::GetDataDevice { id, seat } => match Seat::from_resource(&seat) {
-            Some(seat) => {
-                // ensure the seat user_data is ready
-                seat.user_data()
-                    .insert_if_missing(|| RefCell::new(SeatData::new(log.clone())));
-                let seat_data = seat.user_data().get::<RefCell<SeatData>>().unwrap();
-                let data_device = implement_data_device(
-                    id,
-                    seat.clone(),
-                    callback.clone(),
-                    action_choice.clone(),
-                    log.clone(),
-                );
-                seat_data.borrow_mut().known_devices.push(data_device);
-            }
-            None => {
-                error!(log, "Unmanaged seat given to a data device.");
-            }
-        },
-        _ => unreachable!(),
-    });
-
-    ddm.deref().clone()
+/// Userdata of WlDataOffer
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub enum DataOfferUserData {
```

Comment:
Do we want to expose these implementation details to users?

Reply:
This will go away when we switch to manual object impl

Reply:
Ah I guess I was thinking too far ahead.